carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/domain/Recruitment.kt

```diff
@@ -0,0 +1,109 @@
+package dk.cachet.carp.studies.domain
+
+import dk.cachet.carp.common.UUID
+import dk.cachet.carp.common.ddd.AggregateRoot
+import dk.cachet.carp.common.ddd.DomainEvent
+import dk.cachet.carp.deployment.domain.users.Participation
+import dk.cachet.carp.deployment.domain.users.StudyInvitation
+import dk.cachet.carp.protocols.domain.StudyProtocolSnapshot
+import dk.cachet.carp.studies.domain.users.DeanonymizedParticipation
+import dk.cachet.carp.studies.domain.users.Participant
+
+
+/**
+ * Represents a set of participants recruited for a [Study] identified by [studyId].
+ */
+class Recruitment( val studyId: UUID ) :
+    AggregateRoot<Recruitment, RecruitmentSnapshot, Recruitment.Event>()
+{
+    sealed class Event : DomainEvent()
+    {
+        data class ParticipationAdded( val studyDeploymentId: UUID, val participation: DeanonymizedParticipation ) : Event()
+    }
+
+
+    companion object
+    {
+        fun fromSnapshot( snapshot: RecruitmentSnapshot ): Recruitment
+        {
+            val recruitment = Recruitment( snapshot.studyId )
+            recruitment.creationDate = snapshot.creationDate
+            if ( snapshot.studyProtocol != null && snapshot.invitation != null )
+            {
+                recruitment.readyForDeployment( snapshot.studyProtocol, snapshot.invitation )
+            }
+            for ( p in snapshot.participations )
+            {
+                recruitment._participations[ p.key ] = p.value.toMutableSet()
+            }
+
+            return recruitment
+        }
+    }
+
+
+    /**
+     * The snapshot of the study protocol to which participants in this recruitment can be invited.
+     */
+    var studyProtocol: StudyProtocolSnapshot? = null
+        private set
+
+    /**
+     * The invitation to send to invited participants.
+     */
+    var invitation: StudyInvitation? = null
+        private set
```

Comment:
This data is now duplicated from the `StudyService`. I don't think that is a problem. Within `StudyService` it is mutable, until a `Recruitment` is created within the scope of which it is immutable.

Likely, a similar relation can be made later once we implement the 'pilot' phase in `Study`. During this time, these could be modified, but each modification will likely result in a new corresponding `Recruitment` object. I think this confirms the relevance of having split these two.

Reply:
Looking more at this, I think I can easily refactor this some more to move some of the current ParticipantServiceHost logic as value objects inside of Recruitment. In particular, the logic which extracts master device role names from the protocol and verifies all of them are assigned in AssignParticipantDevices.

All of this is a very good sign. It means that decoupling ParticipantService from StudyService was a good separation of concerns which now allows pushing constraint logic into domain models where it belongs, rather than in application services.